doc/training_and_validation.md

```diff
@@ -0,0 +1,103 @@
+
+```sql
+​```
+SELECT col1, col2, col3
+FROM mytable
+TRAN ...
+​```
+```
+
+The data comes from the standard select part `SELECT col1, col2, col3 FROM mytable`, and let's say the query result looks like the following
+
+| col1   | col2   | col3   |
+| ------ | ------ | ------ |
+| <data> | <data> | <data> |
+| <data> | <data> | <data> |
+| <data> | <data> | <data> |
+| ...    |        |        |
+
+We want to split the result into 80% training data and 20% validation data.
+
+**We add a column sqlflow_random via RAND() and save the result to a temporary table.** Note the `RAND()` function returns a random number between 0 (inclusive) and 1. The result temporary table looks like the following.
+
+| temp_table |        |        |                |
+| ---------- | ------ | ------ | -------------- |
+| col1       | col2   | col3   | sqlflow_random |
+| <data>     | <data> | <data> | 0.3            |
+| <data>     | <data> | <data> | 0.9            |
+| <data>     | <data> | <data> | 0.5            |
+| ...        |        |        |                |
+
+We can generate the corresponding SQL using the following code template
+
+```
+CREATE TABLE {.TempTableName} AS
+    SELECT *, RAND() AS sqlflow_random FROM (
+        {.StandardSQL}
+    )
+```
+
+## How to split
+
+1. Split the temporary dataset
+
+   **We fetch the training/validation data using two different queries respectively.** The query for training data can be written as `SELECT * FROM temp_table WHERE sqlflow_random < 0.8`, which fetches row1 and row3 etc.. The query for validation data can be written as `SELECT * FROM temp_table WHERE sqlflow_random >= 0.8`, which fetches the rest of the rows.
+
+   In SQLFlow, we modify the user-specific dataset to our temp_table restricted to `sqlflow_random >= 0.8` to train a model, then restricted the temp_table  to`sqlflow_random < 0.8` to validate that model. This context is built after the temporary dataset accomplished, passed to `runExtendedSQL`  in `extendedSelect`.
+
+   ```go
+   type extendedSelect struct {
+     // ...
+     training   standardSelect // training dataset
+     validation standardSelect // validation dataset
+   }
+   ```
+
+2. Split `runExtendedSQL` to `train & validate` on training
+
+```go
+func runExtendedSQL(slct string, db *DB, pr *extendedSelect, modelDir string) *PipeReader { 
+  // ...
+  if os.Getenv("SQLFLOW_submitter") == "alps" {
+    pr = split
+    return submitALPS(wr, pr, db, cwd) // training & validation in a same function
+  }
+  
+  if pr.train {
+    train(pr, slct, db, cwd, wr, modelDir)           // using pr.training
+    return validate(pr, slct, db, cwd, wr, modelDir) // using pr.validation
+  }
+  // ...
+}
+```
+
+`validate` returns the evaluation of the trained model.
+
+## Codegen
+
+For tensorflow submitter, we generate `evaluation` code to evaluate the model. 
```

Comment:
For TensorFlow submitter, we generate training dataset and validation dataset according to `extendedSelect.training` and `extendedSelect.validation`.